Recompilation kotlin sources in dev mode should use compiler plugins

**Description**
Using quarkus 0.16.0 in a kotlin project, when hot reloading is triggered it seems like the kotlin files are recompiled without any kotlin compiler plugins. Consider a resource like the following

    @Path("/greeting")
    @ApplicationScoped
    class GreetingResource @Inject constructor(val greetingService: GreetingService) { ... }

In order for this class to work correctly, it needs to not be final and to have a default constructor. Both can be achieved by configuring the maven kotlin-maven-plugin

    <configuration>
        <compilerPlugins>
            <plugin>all-open</plugin>
            <plugin>no-arg</plugin>
        </compilerPlugins>
        <pluginOptions>
            <option>all-open:annotation=javax.enterprise.context.ApplicationScoped</option>
            <option>no-arg:annotation=javax.enterprise.context.ApplicationScoped</option>
        </pluginOptions>
    </configuration>

And indeed the application is working when initially starting with

    mvn compile quarkus:dev

But as soon as the source is modified the reload fails with

    Caused by: javax.enterprise.inject.spi.DeploymentException: Found 2 deployment problems: 
    [1] Normal scoped bean must not be final: CLASS bean [types=[GreetingResource, java.lang.Object], qualifiers=[@Default, @Any], target=GreetingResource]
    [2] Normal scoped beans must declare a non-private constructor with no parameters: CLASS bean [types=[GreetingResource, java.lang.Object], qualifiers=[@Default, @Any], target=GreetingResource]



**Implementation ideas**
Ideally the kotlin extension would reuse the configuration of the kotlin-maven-plugin from pom.xml. Alternatively there would need to be a separate configuration option for plugins of the quarkus-kotlin extension.

Issue #1157 is related, but this issue is actually more generic as there might be more classes or annotations that  need special treatment by the kotlin compiler, which would then break after hot reloading.

--- a/bom/runtime/pom.xml
+++ b/bom/runtime/pom.xml
@@ -2263,6 +2263,11 @@
             </dependency>
             <dependency>
                 <groupId>org.jetbrains.kotlin</groupId>
+                <artifactId>kotlin-compiler</artifactId>
+                <version>${kotlin.version}</version>
+            </dependency>
+            <dependency>
+                <groupId>org.jetbrains.kotlin</groupId>
                 <artifactId>kotlin-stdlib-jdk8</artifactId>
                 <version>${kotlin.version}</version>
             </dependency>

--- a/bom/runtime/pom.xml
+++ b/bom/runtime/pom.xml
@@ -2258,11 +2258,6 @@
 
             <dependency>
                 <groupId>org.jetbrains.kotlin</groupId>
-                <artifactId>kotlin-compiler-embeddable</artifactId>
-                <version>${kotlin.version}</version>
-            </dependency>
-            <dependency>
-                <groupId>org.jetbrains.kotlin</groupId>
                 <artifactId>kotlin-compiler</artifactId>
                 <version>${kotlin.version}</version>
             </dependency>

--- a/build-parent/pom.xml
+++ b/build-parent/pom.xml
@@ -23,7 +23,7 @@
         <!-- These properties are needed in order for them to be resolvable by the generated projects -->
         <!-- Quarkus uses jboss-parent and it comes with 3.8.0-jboss-2, we don't want that in the templates -->
         <compiler-plugin.version>3.8.1</compiler-plugin.version>
-        <kotlin.version>1.3.21</kotlin.version>
+        <kotlin.version>1.3.41</kotlin.version>
         <scala.version>2.12.8</scala.version>
         <scala-plugin.version>4.1.1</scala-plugin.version>
 

--- a/core/devmode/src/main/java/io/quarkus/dev/ClassLoaderCompiler.java
+++ b/core/devmode/src/main/java/io/quarkus/dev/ClassLoaderCompiler.java
@@ -166,7 +166,9 @@ public class ClassLoaderCompiler {
                                     context.getSourceEncoding(),
                                     context.getCompilerOptions(),
                                     context.getSourceJavaVersion(),
-                                    context.getTargetJvmVersion()));
+                                    context.getTargetJvmVersion(),
+                                    context.getCompilerPluginArtifacts(),
+                                    context.getCompilerPluginsOptions()));
                 });
             }
         }

--- a/core/devmode/src/main/java/io/quarkus/dev/CompilationProvider.java
+++ b/core/devmode/src/main/java/io/quarkus/dev/CompilationProvider.java
@@ -32,6 +32,8 @@ public interface CompilationProvider {
         private final List<String> compilerOptions;
         private final String sourceJavaVersion;
         private final String targetJvmVersion;
+        private final List<String> compilePluginArtifacts;
+        private final List<String> compilerPluginOptions;
 
         public Context(
                 String name,
@@ -42,7 +44,9 @@ public interface CompilationProvider {
                 String sourceEncoding,
                 List<String> compilerOptions,
                 String sourceJavaVersion,
-                String targetJvmVersion) {
+                String targetJvmVersion,
+                List<String> compilePluginArtifacts,
+                List<String> compilerPluginOptions) {
 
             this.name = name;
             this.classpath = classpath;
@@ -53,6 +57,8 @@ public interface CompilationProvider {
             this.compilerOptions = compilerOptions == null ? new ArrayList<String>() : compilerOptions;
             this.sourceJavaVersion = sourceJavaVersion;
             this.targetJvmVersion = targetJvmVersion;
+            this.compilePluginArtifacts = compilePluginArtifacts;
+            this.compilerPluginOptions = compilerPluginOptions;
         }
 
         public String getName() {
@@ -89,6 +95,14 @@ public interface CompilationProvider {
 
         public String getTargetJvmVersion() {
             return targetJvmVersion;
+        }
+
+        public List<String> getCompilePluginArtifacts() {
+            return compilePluginArtifacts;
+        }
+
+        public List<String> getCompilerPluginOptions() {
+            return compilerPluginOptions;
         }
     }
 }

--- a/core/devmode/src/main/java/io/quarkus/dev/DevModeContext.java
+++ b/core/devmode/src/main/java/io/quarkus/dev/DevModeContext.java
@@ -36,6 +36,9 @@ public class DevModeContext implements Serializable {
     private String sourceJavaVersion;
     private String targetJvmVersion;
 
+    private List<String> compilerPluginArtifacts;
+    private List<String> compilerPluginsOptions;
+
     public List<URL> getClassPath() {
         return classPath;
     }
@@ -120,6 +123,22 @@ public class DevModeContext implements Serializable {
         this.targetJvmVersion = targetJvmVersion;
     }
 
+    public List<String> getCompilerPluginArtifacts() {
+        return compilerPluginArtifacts;
+    }
+
+    public void setCompilerPluginArtifacts(List<String> compilerPluginArtifacts) {
+        this.compilerPluginArtifacts = compilerPluginArtifacts;
+    }
+
+    public List<String> getCompilerPluginsOptions() {
+        return compilerPluginsOptions;
+    }
+
+    public void setCompilerPluginsOptions(List<String> compilerPluginsOptions) {
+        this.compilerPluginsOptions = compilerPluginsOptions;
+    }
+
     public File getDevModeRunnerJarFile() {
         return devModeRunnerJarFile;
     }
@@ -173,5 +192,4 @@ public class DevModeContext implements Serializable {
             return resourcePath;
         }
     }
-
 }

--- a/devtools/maven/src/main/java/io/quarkus/maven/DevMojo.java
+++ b/devtools/maven/src/main/java/io/quarkus/maven/DevMojo.java
@@ -31,7 +31,11 @@ import java.util.stream.Collectors;
 import java.util.zip.ZipEntry;
 import java.util.zip.ZipOutputStream;
 
+import org.apache.maven.artifact.Artifact;
+import org.apache.maven.artifact.resolver.ArtifactResolutionRequest;
+import org.apache.maven.artifact.resolver.ArtifactResolutionResult;
 import org.apache.maven.execution.MavenSession;
+import org.apache.maven.model.Dependency;
 import org.apache.maven.model.Plugin;
 import org.apache.maven.plugin.AbstractMojo;
 import org.apache.maven.plugin.MojoExecutionException;
@@ -179,6 +183,9 @@ public class DevMojo extends AbstractMojo {
 
     @Component
     private RepositorySystem repoSystem;
+
+    @Component
+    protected org.apache.maven.repository.RepositorySystem system;
 
     @Component
     private Invoker invoker;
@@ -429,6 +436,7 @@ public class DevMojo extends AbstractMojo {
 
     class DevModeRunner {
 
+        private static final String KOTLIN_MAVEN_PLUGIN_GA = "org.jetbrains.kotlin:kotlin-maven-plugin";
         private final List<String> args;
         private Process process;
         private Set<Path> pomFiles = new HashSet<>();
@@ -511,6 +519,8 @@ public class DevMojo extends AbstractMojo {
                     break;
                 }
             }
+
+            setKotlinSpecificFlags(devModeContext);
 
             final AppModel appModel;
             try {
@@ -636,6 +646,50 @@ public class DevMojo extends AbstractMojo {
             args.add("-jar");
             args.add(tempFile.getAbsolutePath());
 
+        }
+
+        private void setKotlinSpecificFlags(DevModeContext devModeContext) {
+            Plugin kotlinMavenPlugin = null;
+            for (Plugin plugin : project.getBuildPlugins()) {
+                if (plugin.getKey().equals(KOTLIN_MAVEN_PLUGIN_GA)) {
+                    kotlinMavenPlugin = plugin;
+                    break;
+                }
+            }
+
+            if (kotlinMavenPlugin == null) {
+                return;
+            }
+
+            getLog().debug("Kotlin Maven plugin detected");
+
+            List<String> compilerPluginArtifacts = new ArrayList<>();
+            List<Dependency> dependencies = kotlinMavenPlugin.getDependencies();
+            for (Dependency dependency : dependencies) {
+                // TODO do we actually need this to resolve the artifact or can we use the repoSystem?
+                Artifact artifact = system.createDependencyArtifact(dependency);
+                ArtifactResolutionResult resolved = system
+                        .resolve(new ArtifactResolutionRequest().setArtifact(artifact));
+
+                if (resolved.getArtifacts().size() != 1) {
+                    getLog().debug(
+                            "Kotlin compiler plugin " + dependency.getArtifactId()
+                                    + " won't be configured for dev-mode because it wasn't configured properly in pom.xml");
+                }
+                Artifact resolvedArtifact = resolved.getArtifacts().iterator().next();
+                compilerPluginArtifacts.add(resolvedArtifact.getFile().toPath().toAbsolutePath().toString());
+            }
+            devModeContext.setCompilerPluginArtifacts(compilerPluginArtifacts);
+
+            List<String> options = new ArrayList<>();
+            Xpp3Dom compilerPluginConfiguration = (Xpp3Dom) kotlinMavenPlugin.getConfiguration();
+            if (compilerPluginConfiguration != null) {
+                Xpp3Dom compilerPluginArgsConfiguration = compilerPluginConfiguration.getChild("pluginOptions");
+                for (Xpp3Dom argConfiguration : compilerPluginArgsConfiguration.getChildren()) {
+                    options.add(argConfiguration.getValue());
+                }
+            }
+            devModeContext.setCompilerPluginsOptions(options);
         }
 
         public Set<Path> getPomFiles() {

--- a/devtools/maven/src/main/java/io/quarkus/maven/DevMojo.java
+++ b/devtools/maven/src/main/java/io/quarkus/maven/DevMojo.java
@@ -31,9 +31,6 @@ import java.util.stream.Collectors;
 import java.util.zip.ZipEntry;
 import java.util.zip.ZipOutputStream;
 
-import org.apache.maven.artifact.Artifact;
-import org.apache.maven.artifact.resolver.ArtifactResolutionRequest;
-import org.apache.maven.artifact.resolver.ArtifactResolutionResult;
 import org.apache.maven.execution.MavenSession;
 import org.apache.maven.model.Dependency;
 import org.apache.maven.model.Plugin;
@@ -54,7 +51,11 @@ import org.apache.maven.toolchain.ToolchainManager;
 import org.codehaus.plexus.util.xml.Xpp3Dom;
 import org.eclipse.aether.RepositorySystem;
 import org.eclipse.aether.RepositorySystemSession;
+import org.eclipse.aether.artifact.DefaultArtifact;
 import org.eclipse.aether.repository.RemoteRepository;
+import org.eclipse.aether.resolution.ArtifactRequest;
+import org.eclipse.aether.resolution.ArtifactResolutionException;
+import org.eclipse.aether.resolution.ArtifactResult;
 
 import io.quarkus.bootstrap.model.AppDependency;
 import io.quarkus.bootstrap.model.AppModel;
@@ -183,9 +184,6 @@ public class DevMojo extends AbstractMojo {
 
     @Component
     private RepositorySystem repoSystem;
-
-    @Component
-    protected org.apache.maven.repository.RepositorySystem system;
 
     @Component
     private Invoker invoker;
@@ -666,18 +664,18 @@ public class DevMojo extends AbstractMojo {
             List<String> compilerPluginArtifacts = new ArrayList<>();
             List<Dependency> dependencies = kotlinMavenPlugin.getDependencies();
             for (Dependency dependency : dependencies) {
-                // TODO do we actually need this to resolve the artifact or can we use the repoSystem?
-                Artifact artifact = system.createDependencyArtifact(dependency);
-                ArtifactResolutionResult resolved = system
-                        .resolve(new ArtifactResolutionRequest().setArtifact(artifact));
+                try {
+                    ArtifactResult resolvedArtifact = repoSystem.resolveArtifact(repoSession,
+                            new ArtifactRequest()
+                                    .setArtifact(new DefaultArtifact(dependency.getGroupId(), dependency.getArtifactId(),
+                                            dependency.getClassifier(), dependency.getType(), dependency.getVersion()))
+                                    .setRepositories(repos));
 
-                if (resolved.getArtifacts().size() != 1) {
-                    getLog().debug(
-                            "Kotlin compiler plugin " + dependency.getArtifactId()
-                                    + " won't be configured for dev-mode because it wasn't configured properly in pom.xml");
+                    compilerPluginArtifacts.add(resolvedArtifact.getArtifact().getFile().toPath().toAbsolutePath().toString());
+                } catch (ArtifactResolutionException e) {
+                    getLog().warn("Unable to properly setup dev-mode for Kotlin", e);
+                    return;
                 }
-                Artifact resolvedArtifact = resolved.getArtifacts().iterator().next();
-                compilerPluginArtifacts.add(resolvedArtifact.getFile().toPath().toAbsolutePath().toString());
             }
             devModeContext.setCompilerPluginArtifacts(compilerPluginArtifacts);
 

--- a/extensions/kotlin/deployment/pom.xml
+++ b/extensions/kotlin/deployment/pom.xml
@@ -31,7 +31,7 @@
 
         <dependency>
             <groupId>org.jetbrains.kotlin</groupId>
-            <artifactId>kotlin-compiler-embeddable</artifactId>
+            <artifactId>kotlin-compiler</artifactId>
         </dependency>
 
     </dependencies>

--- a/extensions/kotlin/deployment/src/main/java/io/quarkus/kotlin/deployment/KotlinCompilationProvider.java
+++ b/extensions/kotlin/deployment/src/main/java/io/quarkus/kotlin/deployment/KotlinCompilationProvider.java
@@ -6,8 +6,11 @@ import java.util.ArrayList;
 import java.util.Collections;
 import java.util.List;
 import java.util.Set;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
 import java.util.stream.Collectors;
 
+import org.jboss.logging.Logger;
 import org.jetbrains.kotlin.cli.common.ExitCode;
 import org.jetbrains.kotlin.cli.common.arguments.K2JVMCompilerArguments;
 import org.jetbrains.kotlin.cli.common.messages.CompilerMessageLocation;
@@ -20,6 +23,12 @@ import io.quarkus.dev.CompilationProvider;
 
 public class KotlinCompilationProvider implements CompilationProvider {
 
+    private static final Logger log = Logger.getLogger(KotlinCompilationProvider.class);
+
+    // see: https://github.com/JetBrains/kotlin/blob/v1.3.41/libraries/tools/kotlin-maven-plugin/src/main/java/org/jetbrains/kotlin/maven/KotlinCompileMojoBase.java#L192
+    private final static Pattern OPTION_PATTERN = Pattern.compile("([^:]+):([^=]+)=(.*)");
+    private static final String KOTLIN_PACKAGE = "org.jetbrains.kotlin";
+
     @Override
     public Set<String> handledExtensions() {
         return Collections.singleton(".kt");
@@ -28,6 +37,30 @@ public class KotlinCompilationProvider implements CompilationProvider {
     @Override
     public void compile(Set<File> filesToCompile, Context context) {
         K2JVMCompilerArguments compilerArguments = new K2JVMCompilerArguments();
+        if (!context.getCompilePluginArtifacts().isEmpty()) {
+            compilerArguments.setPluginClasspaths(context.getCompilePluginArtifacts().toArray(new String[0]));
+        }
+        if (!context.getCompilerPluginOptions().isEmpty()) {
+            List<String> sanitizedOptions = new ArrayList<>(context.getCompilerOptions().size());
+            for (String rawOption : context.getCompilerPluginOptions()) {
+                Matcher matcher = OPTION_PATTERN.matcher(rawOption);
+                if (!matcher.matches()) {
+                    log.warn("Kotlin compiler plugin option " + rawOption + " is invalid");
+                }
+
+                String pluginId = matcher.group(1);
+                if (!pluginId.contains(".")) {
+                    // convert the plugin name to the plugin id by simply removing the dash and adding the kotlin package
+                    // this seems to be the appropriate way of doing things for the plugins that were checked
+                    pluginId = KOTLIN_PACKAGE + "." + pluginId.replace("-", "");
+                }
+                String key = matcher.group(2);
+                String value = matcher.group(3);
+                sanitizedOptions.add("plugin:" + pluginId + ":" + key + "=" + value);
+
+                compilerArguments.setPluginOptions(sanitizedOptions.toArray(new String[0]));
+            }
+        }
         compilerArguments.setClasspath(
                 context.getClasspath().stream().map(File::getAbsolutePath).collect(Collectors.joining(File.pathSeparator)));
         compilerArguments.setDestination(context.getOutputDirectory().getAbsolutePath());
@@ -44,7 +77,7 @@ public class KotlinCompilationProvider implements CompilationProvider {
         }
 
         if (messageCollector.hasErrors()) {
-            throw new RuntimeException("Compilation failed" + String.join("\n", messageCollector.getErrors()));
+            throw new RuntimeException("Compilation failed. " + String.join("\n", messageCollector.getErrors()));
         }
     }
 

--- a/independent-projects/tools/pom.xml
+++ b/independent-projects/tools/pom.xml
@@ -32,7 +32,7 @@
         <!-- These properties are needed in order for them to be resolvable by the generated projects -->
         <!-- Quarkus uses jboss-parent and it comes with 3.8.0-jboss-2, we don't want that in the templates -->
         <compiler-plugin.version>3.8.1</compiler-plugin.version>
-        <kotlin.version>1.3.21</kotlin.version>
+        <kotlin.version>1.3.41</kotlin.version>
         <scala.version>2.12.8</scala.version>
         <scala-plugin.version>4.1.1</scala-plugin.version>
 

--- a/integration-tests/kotlin/src/test/java/io/quarkus/kotlin/maven/it/KotlinDevModeIT.java
+++ b/integration-tests/kotlin/src/test/java/io/quarkus/kotlin/maven/it/KotlinDevModeIT.java
@@ -22,9 +22,9 @@ public class KotlinDevModeIT extends RunAndCheckMojoTestBase {
         runAndCheck();
 
         // Edit the "Hello" message.
-        File source = new File(testDir, "src/main/kotlin/org/acme/HelloResource.kt");
+        File jaxRsResource = new File(testDir, "src/main/kotlin/org/acme/HelloResource.kt");
         String uuid = UUID.randomUUID().toString();
-        filter(source, ImmutableMap.of("return \"hello\"", "return \"" + uuid + "\""));
+        filter(jaxRsResource, ImmutableMap.of("return \"hello\"", "return \"" + uuid + "\""));
 
         // Wait until we get "uuid"
         await()
@@ -34,13 +34,22 @@ public class KotlinDevModeIT extends RunAndCheckMojoTestBase {
         await()
                 .pollDelay(1, TimeUnit.SECONDS)
                 .pollInterval(1, TimeUnit.SECONDS)
-                .until(source::isFile);
+                .until(jaxRsResource::isFile);
 
-        filter(source, ImmutableMap.of(uuid, "carambar"));
+        filter(jaxRsResource, ImmutableMap.of(uuid, "carambar"));
 
         // Wait until we get "carambar"
         await()
                 .pollDelay(1, TimeUnit.SECONDS)
                 .atMost(1, TimeUnit.MINUTES).until(() -> getHttpResponse("/app/hello").contains("carambar"));
+
+        File greetingService = new File(testDir, "src/main/kotlin/org/acme/GreetingService.kt");
+        String newUuid = UUID.randomUUID().toString();
+        filter(greetingService, ImmutableMap.of("\"hello\"", "\"" + newUuid + "\""));
+
+        // Wait until we get "newUuid"
+        await()
+                .pollDelay(1, TimeUnit.SECONDS)
+                .atMost(1, TimeUnit.MINUTES).until(() -> getHttpResponse("/app/hello/bean").contains(newUuid));
     }
 }

--- a/integration-tests/kotlin/src/test/resources/projects/classic-kotlin/pom.xml
+++ b/integration-tests/kotlin/src/test/resources/projects/classic-kotlin/pom.xml
@@ -12,7 +12,7 @@
         <maven.compiler.source>1.8</maven.compiler.source>
         <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
         <maven.compiler.target>1.8</maven.compiler.target>
-        <kotlin.version>1.3.21</kotlin.version>
+        <kotlin.version>1.3.41</kotlin.version>
     </properties>
     <dependencyManagement>
         <dependencies>

--- a/integration-tests/kotlin/src/test/resources/projects/classic-kotlin/src/main/kotlin/org/acme/GreetingService.kt
+++ b/integration-tests/kotlin/src/test/resources/projects/classic-kotlin/src/main/kotlin/org/acme/GreetingService.kt
@@ -1,0 +1,9 @@
+package org.acme
+
+import javax.enterprise.context.ApplicationScoped
+
+@ApplicationScoped
+class GreetingService {
+
+    fun greet() = "hello"
+}

--- a/integration-tests/kotlin/src/test/resources/projects/classic-kotlin/src/main/kotlin/org/acme/HelloResource.kt
+++ b/integration-tests/kotlin/src/test/resources/projects/classic-kotlin/src/main/kotlin/org/acme/HelloResource.kt
@@ -9,7 +9,7 @@ import javax.ws.rs.Produces
 import javax.ws.rs.core.MediaType
 
 @Path("/hello")
-class HelloResource {
+class HelloResource(val greetingService: GreetingService) {
 
     @Inject
     @ConfigProperty(name = "greeting")
@@ -27,4 +27,9 @@ class HelloResource {
     fun greeting(): String? {
         return greeting
     }
+
+    @GET
+    @Path("/bean")
+    @Produces(MediaType.TEXT_PLAIN)
+    fun greetingFromBean()  = greetingService.greet()
 }

